Add remote management design document (#301)
aws/eks-anywhere-packages: design/remote-management.md

```
apiVersion: packages.eks.amazonaws.com/v1alpha1
kind: PackageBundleController
metadata:
  name: billy
  namespace: eksa-packages
spec:
  logLevel: 4
  activeBundle: "v1-21-31"
  upgradeCheckInterval: "24h"
  source:
    registry: public.ecr.aws/eks-anywhere
    repository: eks-anywhere-packages-bundles
```
Currently, the PBC is in the eksa-packages namespace and must be named "eksa-packages-bundle-controller". This will have to be changed to support any name in that namespace.

# PackageBundle (bundle) Custom Resource
We currently set the bundle state to show the user which bundle is active. Since we will have several clusters sharing the same set of bundles, we should remove this state and have the user look at the PBC to determine the active bundle. We may keep bundle state for invalid bundle names for instance.

# Package Custom Resource

> A cluster specific namespace will need to be created for package custom resources in the form "eksa-packages-${cluster_name}". We use the name of the custom resource as the name of the installed package, so this will allow the same name across clusters. For example, to install the hello app on the billy cluster:

Reviewer: I like that this doesn't require us to create namespaces for the customer. I'd prefer to avoid that as much as possible.

> A cluster specific namespace will need to be created for package custom resources in the form "eksa-packages-${cluster_name}". We use the name of the custom resource as the name of the installed package, so this will allow the same name across clusters. For example, to install the hello app on the billy cluster:

Author: I talked with Eric and Lewis about this today and maybe we are okay with the approach as documented here?

```
apiVersion: packages.eks.amazonaws.com/v1alpha1
kind: Package
metadata:
  name: my-hello

>   namespace: eksa-packages-billy

Reviewer: No, sorry, not "the user's name" as in ewwolles, but rather, as in the name for the helm installation given by the user. 

    helm install hello-eks-anywhere --name foo
    
The "foo" above is what I'm referring to.

>   namespace: eksa-packages-billy

Author: The `my-hello` in this example is the user's name. If a user installs a package multiple times on a cluster, they need a unique name for each installation.

>   namespace: eksa-packages-billy

Author: I'm going to push a change to the names. In theory the name of the package could be the same in several namespaces though, but the example is with a unique name

spec:
  packageName: hello-eks-anywhere
```

# Helm Driver
The helm driver will need to be told where to install the package and it will need to set the kubeconfig for the target cluster. In the NewHelm function, we would have to set the `settings.KubeConfig` value to the location of the kubeconfig file. The easiest way to get the kubeconfig file would be to get it on the fly from the secret. CAPI creates a secret for each cluster with the administrative credentials and we can read this file and create a temporary file for the helm install command. There probably is some optimization here.

For example the tlhowe cluster is the management cluster and the billy cluster is the workload cluster:
```
 % k get secret -A | grep kubeconfig
eksa-system                         billy-kubeconfig                                  cluster.x-k8s.io/secret                1      9h
eksa-system                         tlhowe-kubeconfig                                 cluster.x-k8s.io/secret                1      3d6h
```

This [github isse](https://github.com/helm/helm/issues/6910) has some different ideas on how to change the authentication for the helm go client.

> 

Reviewer: Also, if you haven't fixed it yet, there's a typo in "isse" => "issue", and if we're being formal, "github" => "GitHub" (I should review my own for that I guess).

# Helm Chart
The helm chart will need to be changed to require a cluster name during installation. The chart will use the cluster name to create the PBC for the installation and the "eksa-packages-${cluster_name}" namespace for package resources.

# eks-anywhere CLI

> The CLI will need to be changed to pass the cluster name to the helm chart during cluster creation. It will also need cluster name for package generation and imperative package installation.

Reviewer: ```suggestion
The CLI will need to be changed to pass the cluster name to the helm chart during cluster creation. It will also need a cluster name for package generation and imperative package installation.
```

For example the tlhowe cluster is the management cluster and the billy cluster is the workload cluster:
```
 % k get secret -A | grep kubeconfig
eksa-system                         billy-kubeconfig                                  cluster.x-k8s.io/secret                1      9h
eksa-system                         tlhowe-kubeconfig                                 cluster.x-k8s.io/secret                1      3d6h
```

This [github isse](https://github.com/helm/helm/issues/6910) has some different ideas on how to change the authentication for the helm go client.

# Helm Chart
The helm chart will need to be changed to require a cluster name during installation. The chart will use the cluster name to create the PBC for the installation and the "eksa-packages-${cluster_name}" namespace for package resources.

# eks-anywhere CLI
Several changes are required in the CLI:
* pass cluster name to list packages
* pass cluster name to generate package
* during cluster creation create install Package Bundle Controller only if self managed

> * during cluster creation create PackageBundleController custom resource

Author: ```suggestion
* during cluster creation create PackageBundleController custom resource
* possibly deprecate the package controller installation command
* fail the package controller installation if kubeconfig is pointing to a workload cluster
* during the package controller installation command create the PackageBundleController custom resource
```

# PackageBundleController (PBC) Custom Resource
The management cluster will have a PackageBundleController custom resource for each workload cluster. The name of the PBC will be the name of the workload cluster.  This will allow each cluster to have different active bundles, different Kubernetes versions, and potentially different source registries. For example, the PBC for the billy cluster:

```
apiVersion: packages.eks.amazonaws.com/v1alpha1
kind: PackageBundleController
metadata:
  name: billy
  namespace: eksa-packages
spec:
  logLevel: 4
  activeBundle: "v1-21-31"
  upgradeCheckInterval: "24h"
  source:
    registry: public.ecr.aws/eks-anywhere
    repository: eks-anywhere-packages-bundles
```
Currently, the PBC is in the eksa-packages namespace and must be named "eksa-packages-bundle-controller". This will have to be changed to support any cluster name in that namespace.

# PackageBundle (bundle) Custom Resource

> We currently set the bundle state to show the user which bundle is active. Since we will have several clusters sharing the same set of bundles, we should remove this state and have the user look at the PBC to determine the active bundle. We may keep bundle state for invalid bundle names for instance.

Reviewer 2: Not sure if bundles can be shared across clusters as different clusters may have different kubernetes versions. Might be better to have PackageBundles for each cluster keeping the namespace pattern proposed for Package Custom Resource. i.e. "eksa-packages-${cluster_name}"

Author: The PBC determines the active bundle for a particular cluster. There are multiple bundles, but only one `activeBundle` in the cluster specific PBC.

# Package Custom Resource

> A cluster specific namespace will need to be created for package custom resources in the form "eksa-packages-${cluster_name}". We use the name of the custom resource as the name of the installed package, so this will allow the same name across clusters. For example, to install the hello app on the billy cluster:

Reviewer 3: What is the plan for this namespace when a workload cluster is removed? In that case the packages installed there will be uninstalled so we should get rid of the Package resource in this namespace in case the user tries to create a new workload cluster with the same name. Or we could just delete the namespace outright.

Author: When a workload cluster is deleted, the namespace should be deleted. I'll add a sentence about that below

```
apiVersion: packages.eks.amazonaws.com/v1alpha1
kind: Package
metadata:
  name: prod-hello

>   namespace: eksa-packages-prod

Reviewer 3: Based on the above design shouldn't the namespace here be eksa-packages-billy to install on the billy cluster?

Author: ```suggestion
  namespace: eksa-packages-billy
```

spec:
  packageName: hello-eks-anywhere
```

> 

Author: ```suggestion

When a workload cluster is deleted, the namespace associated should be deleted.
```

For example the tlhowe cluster is the management cluster and the billy cluster is the workload cluster:
```
 % k get secret -A | grep kubeconfig
eksa-system                         billy-kubeconfig                                  cluster.x-k8s.io/secret                1      9h
eksa-system                         tlhowe-kubeconfig                                 cluster.x-k8s.io/secret                1      3d6h
```

This [github isse](https://github.com/helm/helm/issues/6910) has some different ideas on how to change the authentication for the helm go client.

# Helm Chart
The helm chart will need to be changed to require a cluster name during installation. The chart will use the cluster name to create the PBC for the installation and the "eksa-packages-${cluster_name}" namespace for package resources.

# eks-anywhere CLI
Several changes are required in the CLI:

> * pass cluster name to list packages

Author: That would only be the case for a self managed cluster, but that is fine with me.

* pass cluster name to generate package
* during cluster creation create install Package Bundle Controller only if self managed
* during cluster creation create PackageBundleController custom resource

> * possibly deprecate the package controller installation command

Reviewer: This makes sense, as I think the package controller is installed by default now. (or will be very very soon)

# Remotely Managing a Cluster with the Package Controller

> The Package Controller should be able to run on the management cluster and install, update and delete packages on a workload cluster. The package custom resources for workload cluster will exist on the management cluster. The package custom resources for workload cluster will exist on the management cluster.

Reviewer: ```suggestion
The Package Controller should be able to run on the management cluster and install, update and delete packages on a workload cluster. The package custom resource for workload clusters will exist on the management cluster.
```
Removed duplication, and tiny tweaking for readability.

# Helm Driver
The helm driver will need to be told where to install the package and it will need to set the kubeconfig for the target cluster. In the NewHelm function, we would have to set the `settings.KubeConfig` value to the location of the kubeconfig file. The easiest way to get the kubeconfig file would be to get it on the fly from the secret. CAPI creates a secret for each cluster with the administrative credentials and we can read this file and create a temporary file for the helm install command. There probably is some optimization here.

For example the tlhowe cluster is the management cluster and the billy cluster is the workload cluster:
```
 % k get secret -A | grep kubeconfig
eksa-system                         billy-kubeconfig                                  cluster.x-k8s.io/secret                1      9h
eksa-system                         tlhowe-kubeconfig                                 cluster.x-k8s.io/secret                1      3d6h
```

This [github isse](https://github.com/helm/helm/issues/6910) has some different ideas on how to change the authentication for the helm go client.

# Helm Chart

> The helm chart will need to be changed to require a cluster name during installation. The chart will use the cluster name to create the PBC for the installation and the "eksa-packages-${cluster_name}" namespace for package resources.

Reviewer: ```suggestion
The Helm chart will need to be changed to require a cluster name during installation. The chart will use the cluster name to create the PBC for the installation and the "eksa-packages-${cluster_name}" namespace for package resources.
```
Nit: we should be capitalizing Helm as well.

# eks-anywhere CLI
Several changes are required in the CLI:
* pass cluster name to list packages
* pass cluster name to generate package
* during cluster creation create install Package Bundle Controller only if self managed
* during cluster creation create PackageBundleController custom resource
* possibly deprecate the package controller installation command
* fail the package controller installation if kubeconfig is pointing to a workload cluster
* during the package controller installation command create the PackageBundleController custom resource
* pass cluster name to imperative installation
* add new command to list PackageBundleControllers

> 

Reviewer 3: Does the CLI need to inform the package bundle controller when a workload cluster is removed? This way we can cleanup the namespaces. Perhaps when deleting workload cluster delete the package bundle controller resource.

Author: I think the CLI should `k delete ns eksa-packages-$clusterName` and `k delete pbc $clusterName` when it is performing a cluster delete

Author: ```suggestion
* do `kubectl delete ns eksa-packages-$clusterName and `kubectl delete pbc $clusterName` when performing a cluster delete
```

Author: Added a line about this
